### PR TITLE
Blog app - integration specs for views and fix n+1 problems

### PR DESCRIPTION
**Pull request by developertariq**
Milestone 
-
- Solved the N+1 problem when fetching all posts and their comments for a user
- Query before solving N+1 problem
Query before solving N+1 problem

- Query after solving N+1 problem 
Query after solving N+1 problem

- Written test cases for User and Post
- User index page:
  - I can see the username of all other users.
  - I can see the profile picture for each user.
  - I can see the number of posts each user has written.
  - When I click on a user, I am redirected to that user's show page.

- user show page:
  - I can see the user's profile picture.
  - I can see the user's username.
  - I can see the number of posts the user has written.
  - I can see the user's bio.
  - I can see the user's first 3 posts.
  - I can see a button that lets me view all of a user's posts.
  - When I click a user's post, it redirects me to that post's show page.
  - When I click to see all posts, it redirects me to the user's post's index page.

- User post index page:
  - I can see the user's profile picture.
  - I can see the user's username.
  - I can see the number of posts the user has written.
  - I can see a post's title.
  - I can see some of the post's body.
  - I can see the first comments on a post.
  - I can see how many comments a post has.
  - I can see how many likes a post has.
  - I can see a section for pagination if there are more posts than fit on the view.
  - When I click on a post, it redirects me to that post's show page.

- Post show page:
  - I can see the post's title.
  - I can see who wrote the post.
  - I can see how many comments it has.
  - I can see how many likes it has.
  - I can see the post body.
  - I can see the username of each commentor.
  - I can see the comment each commentor left.

### CODE REVIEW

File: app/controllers/posts_controller.rb [hunk 1]
```diff
@@ -1,7 +1,11 @@
 class PostsController < ApplicationController
   def index
-    @user = User.find(params[:user_id])
-    @posts = @user.posts.order(created_at: :desc).paginate(page: params[:page], per_page: 2)
+    # @user = User.find(params[:user_id])
+    # @posts = @user.posts.order(created_at: :desc).paginate(page: params[:page], per_page: 2)
```
Comment (DammyShittu) [inline app/controllers/posts_controller.rb]: - [ ] Kindly remove this commented code.

File: app/controllers/posts_controller.rb [hunk 2]
```diff
@@ -1,7 +1,11 @@
...
+    @posts = Post.includes(:author)
+      .where(author_id: params[:user_id])
+      .order(created_at: :desc)
+      .paginate(page: params[:page], per_page: 2)
```
Comment (DammyShittu) [inline app/controllers/posts_controller.rb]: - [ ] Given that the purpose of eliminating this is to reduce the round trips taken to retrieve the related entities of an object, we should be eliminating the number of queries for fetching the **comments** belonging to a post in this `index` action.

  There is also no need to fetch the post author this way. We should first retrieve the author and then fetch the posts belonging to the author along with their comments. Here is a proposed code you can tweak:

  ```
  @user = User.find_by(id: params[:user_id])
  @posts = @user.posts.includes(:comments)
  ...
  ```

**Review by DammyShittu — CHANGES_REQUESTED** [review]
Hi @developertariq and @altontonn,

## Highlights
✔️ Some passing tests

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Required Changes ♻️
- [ ] Some of your tests are failing. Kindly look into them.

  image

_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by mckent05 — APPROVED** [review]
Hi @developertariq, 

# STATUS: APPROVED ✔️ ✔️ 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations and well done! 🎉 
approved


## Highlights
- All linter checks are passing ✔️ ✔️ 
- Your PR is professional 👏 👏 
- All tests are passing and N+1 problem is fixed, well done 👏 

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by developertariq** [comment]
@mckent05 thanks a lot.

**Pull request merged**